csvcubed/csvcubed/constraints/virtualiser.py

```diff
@@ -0,0 +1,196 @@
+                    self.start_end_indicies.append(start_end_index)
+                    last_row = i * MAXIMUM_ROW_SUBSET
+                    i += 1
+
+        # TODO: probably unnessary/OTT, needs some thought.
+        # copy as a precuation ony, I'm concerned about mutabilty
+        # if we're sourcing multiple genrators against the same list,
+        # remove if/once it's been given some thought.
+        subset_yielder = copy.deepcopy(self.start_end_indicies)
+        for start_end_index in subset_yielder:
+            yield start_end_index[0], start_end_index[1]
+
+    def _column_subset_yielder(self, column_label: str) -> pd.DataFrame:
+        """
+        Generator that returns a sinele column dataframe representing a
+        finite subset of the complete dataframe equal to a maximum
+        rowcount of MAXIMUM_ROW_SUBSET
+        """
+
+        first_iteration = True
+
+        for start, end in self._subset_index_yieder():
+            # Keep track of the headers from the first iteraton and reuse them
+            # (names) for later dataframe slices that don't have a header row
+            if first_iteration:
+                one_col_df: pd.DataFrame = pd.read_csv(
+                    self.csv_path, usecols=[column_label], skiprows=start, nrows=end
+                )
+                headers = one_col_df.columns.values
+                first_iteration = False
+            else:
+                one_col_df: pd.DataFrame = pd.read_csv(
+                    self.csv_path,
+                    names=headers,
+                    usecols=[column_label],
+                    skiprows=start,
+                    nrows=end,
+                )
+
+            one_col_df.fillna("", inplace=True)
+            yield one_col_df
+
+    def _column_value_yielder(self, column_label: str) -> str:
+        """
+        A generator that yields every value sequentially from a single
+        column of the source csv.
+
+        Where feasible, you should always seek to use ._unique_column_value_yielder
+        in preference for preformance reasons.
+        """
+
+        for one_col_df in self._column_subset_yielder(column_label):
+            for cell_value in one_col_df[column_label]:
+                yield cell_value
+
+    def _unique_column_value_yielder(self, column_label: str) -> str:
+        """
+        A generator that yields one value taken from a single csv column
+        """
+
+        # If we've done this before, use the cached values
+        if column_label in self.unique_column_value_cache:
+            for cell_value in self.unique_column_value_cache[column_label]:
+                yield cell_value
+
+        else:
+            yielded = []
+
+            one_col_df: pd.DataFrame
+            for one_col_df in self._column_subset_yielder(column_label):
+
+                # new values
+                values_we_havnt_returned = [
+                    x for x in one_col_df[column_label].unique() if x not in yielded
```

Comment:
This would be more performant for large columns if `yielded` were a set rather than a list.